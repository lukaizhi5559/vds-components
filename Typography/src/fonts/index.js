//****** To-do, test loading in the families first asynchronously from check whether to load the url if necessary ******//

let FontConfig = {
  loaded: false,
};

//Turn on or off the font loader
function setFontLoaded(loaded) {
  FontConfig.loaded = loaded ? true : false;
}

function checkFontLoaded() {
  return FontConfig.loaded;
}

// Define and Load the fonts
let customConfig = {
  families: ['Verizon-NHG-eTX', 'Verizon-NHG-eDS'],
};

let fontUri1 = 'https://respframework.verizon.com/verizon-nhg.css';

async function finishLoading() {
  try {
    const WebFont = await import('webfontloader');
    const WebFontConfig = {
      custom: customConfig,
      active: () => setFontLoaded(true),
      inactive: () => console.warn('fonts could not be loaded.'),
    };

    WebFont.load && WebFont.load(WebFontConfig);
  } catch (error) {
    typeof window !== 'undefined' && console.error(error);
  }
}

function checkIfLoaded() {
  if (typeof document === 'undefined') return false;
  const links = Array.from(document.head.getElementsByTagName('link'));
  if (!links || !links.length) return false;

  return links.find(link => {
    return link.getAttribute('href') === fontUri1;
  });
}

function loadFonts() {
  let loaded = checkIfLoaded();

  if (!loaded) {
    customConfig.urls = [fontUri1];
    finishLoading(customConfig);
  } else setFontLoaded(true);
}

if (checkIfLoaded()) setFontLoaded(true);
if (!checkFontLoaded()) loadFonts();

export const Fonts = {
  VerizonNHGeDS: 'Verizon-NHG-eDS', // Display
  VerizonNHGeTX: 'Verizon-NHG-eTX', // Text
  checkFontLoaded: checkFontLoaded,
  setFontLoaded: setFontLoaded,
  loadFonts: loadFonts,
};
