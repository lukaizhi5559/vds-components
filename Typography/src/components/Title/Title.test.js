import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { Title } from '.';
import TypographyConfig from '../../typographyConfig';
import { Fonts } from '../../fonts';
import { TypographyTokens } from '@vds-tokens/typography';
import { ColorTokens } from '@vds-tokens/color';

// Default tests should match corresponding config.js in package for:
//  - typescale = 'MVP'
//  - viewport  = 'desktop'

function _calcSize(remVal) {
  const baseFont = 16;
  let num = parseFloat(remVal);
  return num * baseFont + 'px';
}

function componentToHex(c) {
  var hex = c.toString(16);
  return hex.length == 1 ? '0' + hex : hex;
}

function rgbToHex(rgb) {
  rgb = rgb
    .substr(4)
    .split(')')[0]
    .split(', ');
  return (
    '#' +
    componentToHex(rgb[0]) +
    componentToHex(rgb[1]) +
    componentToHex(rgb[2])
  );
}

// ============== typescale - MVP ============== //
//                                               //
// ============================================= //

test('should render title default props correctly', () => {
  const content = 'Title default';
  const { getByText, container } = render(<Title>{content}</Title>);
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  fireEvent.click(container.firstChild);

  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    // expect(styles['font-weight']).toBe('750');
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});

test('should render title custom props correctly', () => {
  const content = 'Title default';
  const { getByText, container } = render(
    <Title size={'medium'} fontSize={'20px'} lineHeight={'24px'}>
      {content}
    </Title>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  fireEvent.click(container.firstChild);

  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    // expect(styles['font-weight']).toBe('750');
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});

// Default - w/ typescale="MVP"
test('should render title MVP typescale correctly', () => {
  const content = 'Title default';
  const { getByText, container } = render(
    <Title typescale={'MVP'}>{content}</Title>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});
// Default - w/ theme="{typescale: "MVP"}
test('should render title theme props correctly', () => {
  const content = 'Title MVP w/ theme';
  const theme = { typescale: 'MVP' };
  const { getByText, container } = render(
    <Title theme={theme}>{content}</Title>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});

//Test passing a different primitive
test('should render title primitive as h2', () => {
  const content = 'Title h2';
  const { getByText, container } = render(
    <Title primitive={'h2'}>{content}</Title>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  const elem = container.querySelector('h2');

  expect(elem).toBeInTheDocument();
});

// Test passing a different size
test('should render title large correctly', () => {
  const content = 'Title large';
  const { getByText, container } = render(
    <Title size={'large'}>{content}</Title>
  );

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[32].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[36].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.light.value
    );
  }
});

// Test passing a different color
test('should render title grey faded correctly', () => {
  const content = 'Title grey faded';
  const { getByText, container } = render(
    <Title color={'#F6F6F6'}>{content}</Title>
  );

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(styles['color']).toBe('rgb(246, 246, 246)');
  }
});

// Test changing the viewport
test('should render title with default styles on mobile correctly', () => {
  const content = 'Title on mobile';
  const { getByText, container } = render(
    <Title viewport={'mobile'}>{content}</Title>
  );

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);
  //Find styling props
  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[16].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[20].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
  }
});
test('should render title medium with bold prop true', () => {
  const content = 'Title medium normal';
  const { getByText, container } = render(
    <Title
      id="test"
      aria-labelledby="test"
      size="medium"
      bold={true}
      viewport="desktop"
    >
      {content}
    </Title>
  );

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[24].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[28].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.bold.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});

test('should render title bold prop false only when size medium', () => {
  const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
  const content = 'Title large';
  const { getByText, container } = render(
    <Title
      id="test"
      aria-labelledby="test"
      size="large"
      bold={false}
      viewport="desktop"
    >
      {content}
    </Title>
  );

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[32].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[36].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.light.value
    );
  }
  warn.mockReset();
});

// ============== typescale - VDS ============== //
//                                               //
// ============================================= //

// Default - w/ typescale="VDS"
test('should render title VDS typescale correctly', () => {
  const content = 'Title VDS typescale';
  const { getByText, container } = render(
    <Title typescale={'VDS'}>{content}</Title>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});
// Default - w/ theme="{typescale: 'VDS'}"
test('should render title VDS theme correctly', () => {
  const content = 'Title VDS theme';
  const theme = { typescale: 'VDS' };
  const { getByText, container } = render(
    <Title theme={theme}>{content}</Title>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
});
// Default - w/ settypescale
test('should render title VDS typescale correctly', () => {
  TypographyConfig.settypescale('VDS');
  const content = 'Title VDS settypescale';
  const { getByText, container } = render(
    <Title id="test" aria-labelledby="test">
      {content}
    </Title>
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(content).textContent).toBe(content);

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
  TypographyConfig.settypescale('MVP');
});

// Default - w/ settypescale
test('should render title small', () => {
  TypographyConfig.settypescale('VDS');
  const content = 'Title VDS settypescale';
  const { getByText, container } = render(
    <Title id="test" aria-labelledby="test" viewport="desktop">
      {content}
    </Title>
  );

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[20].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[24].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
  TypographyConfig.settypescale('MVP');
});

// Default - w/ settypescale
test('should render title medium with bold prop false', () => {
  TypographyConfig.settypescale('VDS');
  const content = 'Title VDS medium normal';
  const { getByText, container } = render(
    <Title
      id="test"
      aria-labelledby="test"
      size="medium"
      bold={false}
      viewport="desktop"
    >
      {content}
    </Title>
  );

  const styles = window.getComputedStyle(container.firstChild)._values;
  if (styles) {
    expect(_calcSize(styles['font-size'])).toBe(
      TypographyTokens.fontsize.title[24].value
    );
    expect(_calcSize(styles['line-height'])).toBe(
      TypographyTokens.lineheight.title[28].value
    );
    expect(parseFloat(styles['font-weight'])).toBe(
      TypographyTokens.fontweight.regular.value
    );
    expect(styles['font-family']).toBe(
      'Verizon-NHG-eDS,Helvetica,Arial,Sans-serif'
    );
    expect(rgbToHex(styles.color)).toBe(ColorTokens.palette.black.value);
  }
  TypographyConfig.settypescale('MVP');
});
