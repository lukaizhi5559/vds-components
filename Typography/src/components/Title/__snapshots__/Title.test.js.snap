// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render title MVP typescale correctly 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
>
  Title default
</h5>
`;

exports[`should render title VDS theme correctly 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
>
  Title VDS theme
</h5>
`;

exports[`should render title VDS typescale correctly 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
>
  Title VDS typescale
</h5>
`;

exports[`should render title VDS typescale correctly 2`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
  id="test"
>
  Title VDS settypescale
</h5>
`;

exports[`should render title custom props correctly 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h4
  aria-hidden="false"
  class="c0 c1"
>
  Title default
</h4>
`;

exports[`should render title default props correctly 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
>
  Title default
</h5>
`;

exports[`should render title grey faded correctly 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #F6F6F6;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
>
  Title grey faded
</h5>
`;

exports[`should render title large correctly 1`] = `
.c1 {
  font-size: 2rem;
  font-weight: 300;
  line-height: 2.25rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h3
  aria-hidden="false"
  class="c0 c1"
>
  Title large
</h3>
`;

exports[`should render title primitive as h2 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h2
  aria-hidden="false"
  class="c0 c1"
>
  Title h2
</h2>
`;

exports[`should render title theme props correctly 1`] = `
.c1 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.5rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
>
  Title MVP w/ theme
</h5>
`;

exports[`should render title with default styles on mobile correctly 1`] = `
.c1 {
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.25rem;
  font-family: Verizon-NHG-eDS,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:active,
.c0:focus,
.c0:hover {
  outline: none;
}

<h5
  aria-hidden="false"
  class="c0 c1"
>
  Title on mobile
</h5>
`;
